bluemira/radiation_transport/radiation_profile.py

```diff
@@ -0,0 +1,232 @@
+        rho_core2 = np.linspace(0.95 * self.rho_ped, self.rho_ped)
+        rho_core = np.append(rho_core1, rho_core2)
+
+        # Plasma mantle for rho_core < rho < 1
+        rho_sep = np.linspace(self.rho_ped, 1)
+
+        self.rho = np.append(rho_core, rho_sep)
+
+    def calculate_electron_density_and_temperature(self):
+        """
+        Calculation of electron density and electron temperature.
+        The region that extends through the plasma core until its
+        outer layer, named pedestal, is referred as "interior".
+        The region that extends from the pedestal to the separatrix
+        is referred as "exterior".
+
+        Returns
+        -------
+        ne: array
+        te: array
+        """
+        i_interior = np.where((self.rho >= 0) & (self.rho <= self.rho_ped))[0]
+
+        n_grad_ped0 = self.params["ne0"] - self.params["neped"]
+        t_grad_ped0 = self.params["te0"] - self.params["teped"]
+
+        rho_ratio_n = (
+            1 - (self.rho[i_interior] ** 2 / self.rho_ped ** 2)
+        ) ** self.params["alphan"]
+        rho_ratio_t = (
+            1
+            - (
+                self.rho[i_interior] ** self.params["tbeta"]
+                / self.rho_ped ** self.params["tbeta"]
+            )
+        ) ** self.params["alphat"]
+
+        ne_i = self.params["neped"] + n_grad_ped0 * rho_ratio_n
+        te_i = self.params["teped"] + t_grad_ped0 * rho_ratio_t
+
+        i_exterior = np.where((self.rho > self.rho_ped) & (self.rho <= 1))[0]
+
+        n_grad_sepped = self.params["neped"] - self.params["nesep"]
+        t_grad_sepped = self.params["teped"] - self.params["tesep"]
+
+        rho_ratio = (1 - self.rho[i_exterior]) / (1 - self.rho_ped)
+
+        ne_e = self.params["nesep"] + n_grad_sepped * rho_ratio
+        te_e = self.params["tesep"] + t_grad_sepped * rho_ratio
+
+        ne = np.append(ne_i, ne_e)
+        te = np.append(te_i, te_e)
+
+        return ne, te
+
+    def calculate_z_species(self, species_file, fimp_no, te):
+        """
+        Calculation of species ion charge, in condition of quasi-neutrality.
+
+        Parameters
+        ----------
+        species_file: string
+            impurity data file (.dat)
+            currently located in process/process/data/impuritydata (__Lzdata.dat)
+            e.g. 1 letter acronym: Hydrogen -> H_Lzdata.dat
+            e.g. 2 letters acronym: Helium -> HeLzdata.dat
+        fimp_no: string
+            impurity fraction to find in MFILE.dat: fimp(no -> no = [01,14]
+        te: array
+            electron temperature
+
+        Returns
+        -------
+        species_frac*z_val**2: np.array
+            species ion charge
+        """
+        species_data = Path(Path(imp_data.__file__).parent, species_file)
+        species_frac = self.process_solver.get_process_parameters(fimp_no)
+
+        t_ref = np.genfromtxt(species_data).T[0]
+        z_ref = np.genfromtxt(species_data).T[2]
```

Comment:
Probably dont want to read in the same file twice (don't forget the last comma otherwise you wont get what you want
```suggestion
        t_ref, z_ref = np.genfromtxt(species_data).T[(0,2),]
```